drivers/gpio/gpio_tca9538.c

```diff
@@ -0,0 +1,541 @@
+{
+	const struct tca9538_config *cfg = dev->config;
+	struct tca9538_drv_data *drv_data = dev->data;
+
+	if (cfg->interrupt_enabled) {
+		/* Device does not support level-triggered interrupts. */
+		if (mode == GPIO_INT_MODE_LEVEL) {
+			return -ENOTSUP;
+		}
+
+		struct tca9538_irq_state *irq = &drv_data->irq_state;
+
+		k_sem_take(&drv_data->lock, K_FOREVER);
+
+		if (mode == GPIO_INT_MODE_DISABLED) {
+			drv_data->cb_pins &= ~BIT(pin);
+			irq->mask |= BIT(pin);
+		} else { /* GPIO_INT_MODE_EDGE */
+			drv_data->cb_pins |= BIT(pin);
+			irq->mask &= ~BIT(pin);
+			if (trig == GPIO_INT_TRIG_BOTH) {
+				irq->falling |= BIT(pin);
+				irq->rising |= BIT(pin);
+			} else if (trig == GPIO_INT_TRIG_LOW) {
+				irq->falling |= BIT(pin);
+				irq->rising &= ~BIT(pin);
+			} else if (trig == GPIO_INT_TRIG_HIGH) {
+				irq->falling &= ~BIT(pin);
+				irq->rising |= BIT(pin);
+			}
+		}
+		k_sem_give(&drv_data->lock);
+	}
+
+	return 0;
+}
+
+static int gpio_tca9538_manage_callback(const struct device *dev,
+					struct gpio_callback *callback,
+					bool set)
+{
+	struct tca9538_drv_data *data = dev->data;
+
+	return gpio_manage_callback(&data->cb, callback, set);
+}
+
+/**
+ * @brief Initialization function of TCA9538
+ *
+ * This sets initial input/ output configuration, output states and
+ * input polarity. The interrupt is configured if this is enabled.
+ *
+ * @param dev Device struct
+ * @return 0 if successful, failed otherwise.
+ */
+static int gpio_tca9538_init(const struct device *dev)
+{
+	const struct tca9538_config *cfg = dev->config;
+	struct tca9538_drv_data *drv_data = dev->data;
+	int rc = 0;
+	char tx_dat[2];
+
+	if (!device_is_ready(cfg->i2c_dev)) {
+		LOG_ERR("I2C device not found");
+		return -EINVAL;
+	}
+
+	if (cfg->interrupt_enabled) {
+		if (!device_is_ready(cfg->gpio_int.port)) {
+			LOG_ERR("Cannot get pointer to gpio interrupt device");
+			rc = -EINVAL;
+		} else {
+			drv_data->dev = dev;
+
+			k_work_init(&drv_data->work, gpio_tca9538_work_handler);
+
+			gpio_pin_configure(cfg->gpio_int.port, cfg->gpio_int.pin,
+					   GPIO_INPUT | cfg->gpio_int.dt_flags);
+			gpio_pin_interrupt_configure(cfg->gpio_int.port,
+						     cfg->gpio_int.pin,
+						     GPIO_INT_EDGE_TO_ACTIVE);
+
+			gpio_init_callback(&drv_data->gpio_cb,
+					   gpio_tca9538_init_cb,
+					   BIT(cfg->gpio_int.pin));
+
+			gpio_add_callback(cfg->gpio_int.port,
+					  &drv_data->gpio_cb);
+
+			drv_data->irq_state = (struct tca9538_irq_state){
+				.mask = ALL_PINS,
+			};
+		}
+	}
+
+	if (rc == 0) {
+		tx_dat[0] = TCA9538_OUTPUT_PORT;
+		tx_dat[1] = drv_data->pin_state.output;
+		rc = i2c_write(cfg->i2c_dev, tx_dat, sizeof(tx_dat),
+			       cfg->i2c_addr);
+	}
+
+	if (rc == 0) {
+		tx_dat[0] = TCA9538_CONFIGURATION;
+		tx_dat[1] = drv_data->pin_state.dir;
+		rc = i2c_write(cfg->i2c_dev, tx_dat, sizeof(tx_dat),
+			       cfg->i2c_addr);
+	}
+
+	/* Do an initial read, this also clears the interrupt pin */
+	if (rc == 0) {
+		rc = update_input(dev);
```

Comment:
We need to initialize full `drv_data->pin_state` struct, not only the `input` field.

Reply:
Moved to the DTS initialisation section.